Fix flows undo/redo

epaminond opened the pull request:
I've spit snapshots history into 2 stacks + `currentSnapshot` and things seem to work fine now.
@slvnperron , should we use Stack from `immutable` as in DraftJS, or shall we stick with plain Arrays (as I did for now)?

epaminond committed:
Fix flows undo/redo (resolve #371)

emirotin reviewed (approved):
lgtm
I wouldn't use Immutable without much need if things work nicely as they are with native arrays.

slvnperron commented:
The code is #1, but the same problem same to remain. Please see the GIF attached here, there's two issues that I show:

(1) Is that on a fresh page load, there's an "undo" record. We haven't modified anything manually so undo should be disabled (i.e. undoStack should be empty on flow loading)

(2) For some reason I think that we are recording the snapshots too granularly. I am not sure what the best thing is here, but perhaps we should use a [lodash debounce](https://lodash.com/docs/4.17.4) with `leading=false`, `trailing=true` at like `2000ms`, so that we only record a snapshot when state is done changing.

It seems like a faster fix (both CPU-wise and coding-wise). Does that make sense? @epaminond @emirotin 

![undo_redo](https://user-images.githubusercontent.com/1315508/35712001-ed7606a8-078d-11e8-8e79-7d5f38f5f92f.gif)

emirotin commented:
If we have complex requirements maybe we should just go for 3rd party solution? IMO we should simply ignore transient events (drag) and record on commit events (drop). Similar to how `redux-storage` lets you define on which events to serialize state and on which not.

slvnperron commented:
@emirotin That is currently what we do but seems like we capture too many events

@epaminond I think we should give `debounce` a try here since it's very quick to test. We also need to refresh the graph after a undo/redo so that links are properly wired

epaminond commented:
@slvnperron , an issue with redundant "undo" records comes from `refreshFlowsLinks` action that triggers update just to make links refresh. Checking how this can be fixed.

slvnperron commented:
@epaminond cool thanks. We have a function that checks for the difference in the flow and if there's no difference it should skip creating a record. Perhaps my serialization logic is broken?

epaminond commented:
@slvnperron , the problem was that there was difference. E.g. we may sometimes call `updateFlowNode` and then immediately `updateFlow`. I've refactored undo-redo logic and changed it a little bit to fix issues you pointed out. Please check whether it's now closer to what you expected.

slvnperron commented:
@epaminond This works MUCH better now, great job!! 🎉 

PS: Weird it says the last commit dates from 3 days ago on my side and GitHub shows only 1 commit, not sure what happened? I also got a merge conflict trying to pull in your changes 😕

epaminond commented:
@slvnperron , that's probably because I used `git commit --amend` to keep everything in one commit.

slvnperron commented:
Ah! Didn't know that even existed. Thanks

On Feb 4, 2018 4:04 AM, "Alex Pletnov" <notifications@github.com> wrote:

> @slvnperron <https://github.com/slvnperron> , that's probably because I
> used git commit --amend to keep everything in one commit.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/botpress/botpress/pull/420#issuecomment-362891856>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABQStCHH5EyCjJ_ezgZ7XJsnWhcYSBUOks5tRXK6gaJpZM4R1O_y>
> .
>
